### PR TITLE
Issue 34 Create LangChain workflow and implement memory

### PR DESCRIPTION
**Pull request by sgt-ppepper**
## Added basic AI Travel Assistant logic with memory in LangChain

- **Created custom `SummaryChatMessageHistory` class**  
  A wrapper for `ConversationSummaryMemory` to support its usage in `RunnableWithMessageHistory` (not supported out-of-the-box, it has to inherit `BaseChatMessageHistory`).   

- **Added `itinerary_chain`**  
  - Uses the **Groq LLaMA3** model via LangChain to generate structured travel itineraries.  
  - Has a clear response format (days → morning/afternoon/evening, activity type, transportation, budget).  
  - Integrated memory to store context and take previous messages into account.  

- **Streaming responses**  
  Supports both streaming output (`stream`) and retrieving the full response (`invoke`).  

**Result**: The assistant can maintain continuous conversation with the user, remember previous requests and itineraries, and provide detailed, structured travel recommendations based on the conversation history.

### CODE REVIEW

File: app/chains/itinerary_chain.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+runnable_with_history = RunnableWithMessageHistory(
+    runnable=chain,
+    # Always return the same memory object for session history
+    # NOTE: If we need to handle multiple sessions, we can modify this to return different memory instances based on session_id
+    get_session_history=lambda session_id: memory,
+    input_messages_key="user_input",
+    history_messages_key="chat_history"
+)
+
+async def stream_response(user_input, session_id="default_session"):
```
[the remark this answers is not in the document]
Comment (sgt-ppepper) [reply app/chains/itinerary_chain.py:106]: It`s a reasonable comment. Actually asynchronous approach was used mostly for future, because as I know it`s best practice for agents development. But for now synchronous version is more than enough. So, if this point will be  supported by other reviewers I will edit the code to use synchronous version of the code.

File: app/chains/itinerary_chain.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
+from memory.async_summary import AsyncConversationSummaryMemory
```
Comment (pashamagpie) [inline app/chains/itinerary_chain.py:1]: nit: you can run linter to ensure code style consistency (refer to section [2.3.4 Linting] in dev-setup.md)

**Review by pashamagpie — APPROVED** [review]
approved

File: app/chains/itinerary_chain.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    response = ""
+    try:
+        result = runnable_with_history.invoke(
+            {"user_input": user_input},
+            config={"configurable": {"session_id": session_id}}
+        )
+        response = result.content if hasattr(result, 'content') else str(result)
+        print(response)
+    except Exception as e:
+        print(f"\nERROR: {e}")
```
Comment (mehalyna) [inline app/chains/itinerary_chain.py:90]: Same as above regarding exception handling and printing - use logging and structured error reporting.

File: app/memory/custom_summary_memory.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+from langchain.memory import ConversationSummaryMemory
+
+class SummaryChatMessageHistory(BaseChatMessageHistory):
+    def __init__(self, summary_memory: ConversationSummaryMemory):
+        self.summary_memory = summary_memory
+
+    @property
+    def messages(self) -> list[BaseMessage]:
+        if self.summary_memory.buffer:
+            return [SystemMessage(content=self.summary_memory.buffer)]
```
Comment (mehalyna) [inline app/memory/custom_summary_memory.py:12]: Returning just a SystemMessage containing the summary may not be compatible with consumers expecting a full message history. Add documentation explaining this design choice, or consider including the full chat history if needed.

File: app/memory/custom_summary_memory.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        return []
+
+    def add_message(self, message: BaseMessage) -> None:
+        self.summary_memory.chat_memory.add_message(message)
```
Comment (mehalyna) [inline app/memory/custom_summary_memory.py:16]: Add error handling to verify that message is the correct type (BaseMessage or subclass). Unexpected message formats could cause errors downstream.

File: app/memory/custom_summary_memory.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        if len(self.summary_memory.chat_memory.messages) == 2:  # Після додавання human та ai
```
Comment (mehalyna) [inline app/memory/custom_summary_memory.py:17]: The hardcoded “2” assumes a strict alternation of human and AI messages. Explain this assumption in a comment or make it a configurable parameter, as some interactions may involve more messages before summary.

File: app/memory/custom_summary_memory.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+            self.summary_memory.buffer = self.summary_memory.predict_new_summary(
+                self.summary_memory.chat_memory.messages,
+                self.summary_memory.buffer
+            )
```
Comment (mehalyna) [inline app/memory/custom_summary_memory.py:21]: Add try/except to handle any exceptions from predict_new_summary.
Consider passing only new messages (not all messages) if the summary model expects incremental updates.

File: app/memory/custom_summary_memory.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+            self.summary_memory.chat_memory.clear()
```
Comment (mehalyna) [inline app/memory/custom_summary_memory.py:22]: Before clearing, ensure that the buffer is updated successfully to avoid losing unsummarized messages.
Add a comment clarifying why clearing is safe at this stage.

File: app/memory/custom_summary_memory.py
Comment (mehalyna) [inline app/memory/custom_summary_memory.py:1]: - Add docstrings for the class and each method to clarify their purpose and expected behavior.
- Add logging (not just print) for important events such as summary update, buffer clearing, or exceptions.
- If this class could be used in a multi-threaded environment (e.g., serving concurrent requests), consider thread-safety for the buffer and chat_memory.
- Provide unit tests, especially for edge cases where the buffer is empty, or when messages arrive out-of-order.

File: app/utils/read_prompt_from_file.py
Comment (mehalyna) [inline app/utils/read_prompt_from_file.py:1]: Consider adding a module-level docstring at the top of the file to briefly describe the purpose of this utility.

Comment (mehalyna) [inline app/utils/read_prompt_from_file.py:1]: You may want to add unit tests to verify this function's behavior, especially for edge cases like non-existent files or permission errors.
Consider type-checking or validating the file_path argument to ensure it is a string and not empty.

File: app/chains/itinerary_chain.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+chain = RunnablePassthrough.assign(
+    chat_history=lambda x: x["chat_history"][0].content if x["chat_history"] else ""
+) | prompt | llm
+
+# Wrapper for message history
+runnable_with_history = RunnableWithMessageHistory(
+    runnable=chain,
+    # Always return the same memory object for session history
+    # NOTE: If we need to handle multiple sessions, we can modify this to return different memory instances based on session_id
+    get_session_history=lambda session_id: SummaryChatMessageHistory(memory),
```
[the remark this answers is not in the document]
Comment (sgt-ppepper) [reply app/chains/itinerary_chain.py:92]: At this point we will stick to 1 global session. Supporting of multiple sessions will added in the future

File: app/chains/itinerary_chain.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+except Exception as e:
+    print(f"ERROR loading prompts: {e}")
+    exit(1)
+
+prompt = PromptTemplate(
+    input_variables=["chat_history", "user_input"],
+    template=itinerary_template
+)
+
+memory = ConversationSummaryMemory(llm=llm)
```
[the remark this answers is not in the document]
Comment (sgt-ppepper) [reply app/chains/itinerary_chain.py]: At this point we will stick to 1 global session. Supporting of multiple sessions will added in the future

File: app/chains/itinerary_chain.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+memory.prompt = PromptTemplate(
+    input_variables=["summary", "new_lines"],
+    template=summary_template
+)
```
[the remark this answers is not in the document]
Comment (sgt-ppepper) [reply app/chains/itinerary_chain.py]: done

File: app/chains/itinerary_chain.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sgt-ppepper) [reply app/chains/itinerary_chain.py]: logging done

Comment (sgt-ppepper) [reply app/chains/itinerary_chain.py]: still using print(), logging done

File: app/chains/itinerary_chain.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+def main():
+    """
+    Main function to run the assistant
+    """
+    print("\n\nHey! I am your travel assistant. How can I help?")
+    
+    session_id = "default_session"
+    while True:
+        user_input = input("\nQuery ('q' to quit): ")
```
[the remark this answers is not in the document]
Comment (sgt-ppepper) [reply app/chains/itinerary_chain.py]: done

File: app/chains/itinerary_chain.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+        # Streaming response
+        stream_response(user_input, session_id)
+        
+        # Response without streaming
+        #full_response(user_input, session_id)
+
+        # Debugging: Print current memory state
+        # Uncomment the line below to see the memory state after each query
+        print(f"\n\nMemory: {memory.buffer}")
```
[the remark this answers is not in the document]
Comment (sgt-ppepper) [reply app/chains/itinerary_chain.py]: commented ar this point

File: app/memory/custom_summary_memory.py (hunk 7 above)
Comment (sgt-ppepper) [reply app/memory/custom_summary_memory.py]: done

File: app/utils/read_prompt_from_file.py
Comment (sgt-ppepper) [reply app/utils/read_prompt_from_file.py:1]: done

Comment (sgt-ppepper) [reply app/utils/read_prompt_from_file.py:1]: done

File: app/memory/custom_summary_memory.py
Comment (sgt-ppepper) [reply app/memory/custom_summary_memory.py:1]: done

File: app/memory/custom_summary_memory.py (hunk 8 above)
Comment (sgt-ppepper) [reply app/memory/custom_summary_memory.py]: done

File: app/memory/custom_summary_memory.py (hunk 4 above)
Comment (sgt-ppepper) [reply app/memory/custom_summary_memory.py:64]: The point is we don`t store full message history, only summary - because we are using ConversationSummaryMemory

File: app/chains/itinerary_chain.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+from langchain_groq import ChatGroq
+from langchain.prompts import PromptTemplate
+from langchain_core.runnables.history import RunnableWithMessageHistory
+import os
+import time
+from app.utils.read_prompt_from_file import load_prompt_from_file
+from app.memory.custom_summary_memory import SummaryChatMessageHistory
+from dotenv import load_dotenv
+
+load_dotenv()  # Load environment variables from .env file
```
[the remark this answers is not in the document]
Comment (sgt-ppepper) [reply app/chains/itinerary_chain.py:19]: not quite understood your point, would be glad for future discussion on the meeting

**Review by Roman-Terendiy — APPROVED** [review]
